protocol/protocol.go

```diff
@@ -0,0 +1,687 @@
+		return mjd.Mjd{}, 0, errors.New("protocol: invalid delegation signature")
+	}
+
+	delegation, err := Decode(delegationBytes)
+	if err != nil {
+		return mjd.Mjd{}, 0, errors.New("protocol: failed to parse delegation: " + err.Error())
+	}
+
+	minTime, err := getTimestamp(delegation, tagMINT, "minimum time")
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	maxTime, err := getTimestamp(delegation, tagMAXT, "maximum time")
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	delegatedPublicKey, err := getFixedLength(delegation, tagPUBK, "public key", ed25519.PublicKeySize)
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	responseSigBytes, err := getFixedLength(reply, tagSIG, "signature", ed25519.SignatureSize)
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	signedResponseBytes, ok := reply[tagSREP]
+	if !ok {
+		return mjd.Mjd{}, 0, errors.New("protocol: response is missing signed portion")
+	}
+
+	if !ed25519.Verify(delegatedPublicKey, []byte(signedResponseContext+string(signedResponseBytes)), responseSigBytes) {
+		return mjd.Mjd{}, 0, errors.New("protocol: invalid response signature")
+	}
+
+	signedResponse, err := Decode(signedResponseBytes)
+	if err != nil {
+		return mjd.Mjd{}, 0, errors.New("protocol: failed to parse signed response: " + err.Error())
+	}
+
+	root, err := getFixedLength(signedResponse, tagROOT, "root", 32)
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	midpoint, err := getTimestamp(signedResponse, tagMIDP, "midpoint")
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	radius, err = getUint32(signedResponse, tagRADI, "radius")
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	// We now need to do some arithmetic.
+
+	if maxTime.Cmp(minTime) < 0 {
+		return mjd.Mjd{}, 0, errors.New("protocol: invalid delegation range")
+	}
+
+	if midpoint.Cmp(minTime) < 0 || maxTime.Cmp(midpoint) < 0 {
+		return mjd.Mjd{}, 0, errors.New("protocol: timestamp out of range for delegation")
+	}
+
+	index, err := getUint32(reply, tagINDX, "index")
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	path, err := getValue(reply, tagPATH, "path")
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+	if len(path)%32 != 0 {
+		return mjd.Mjd{}, 0, errors.New("protocol: path is not a multiple of 32")
+	}
+
+	var hash [sha512.Size]byte
+	hashLeaf(&hash, nonce[:])
+
+	for len(path) > 0 {
+		pathElementIsRight := index&1 == 0
+		if pathElementIsRight {
+			hashNode(&hash, hash[:32], path[:32])
+		} else {
+			hashNode(&hash, path[:32], hash[:32])
+		}
+
+		index >>= 1
+		path = path[32:]
+	}
+
+	if !bytes.Equal(hash[:32], root) {
+		return mjd.Mjd{}, 0, errors.New("protocol: calculated tree root doesn't match signed root")
+	}
+
+	return midpoint, radius, nil
+}
+
+// EncapsulatePacket creates a UDP/TCP payload containing a roughtime message.
+func EncapsulatePacket(version uint32, message []byte) []byte {
+	length := len(message)
+	ret := make([]byte, length+12)
+	copy(ret, "ROUGHTIM")
```

Comment:
You can create a tag variable for `ROUGHTIM`.

Reply:
This isn't a tag, just a header.